File: shells/android/java/dev/angularnative/AnBuiltinModules.java
package dev.angularnative;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;

import org.json.JSONObject;

import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;

/**
 * The built-in modules of the Android shell.
 *
 * <p>It is the twin of {@link AnPluginRegistry} with one difference that matters: who is inside is
 * not read from anybody's {@code package.json}. These modules ship with the framework and every host
 * has them, so the list is written here and in {@code BUILTIN_MODULES} on the Rust side, and {@code
 * scripts/check-builtins.sh} reads both.
 *
 * <p>It also owns the request codes. A module that has to open a system chooser needs one, and two
 * modules picking the same number by hand would each get the other's answer.
 */
public final class AnBuiltinModules {

    private static final String TAG = "angular-native";

    /**
     * Where this shell's request codes start. Far enough from 0 that an app embedding the runtime
     * and starting activities of its own does not collide with it by accident.
     */
    private static final int FIRST_REQUEST_CODE = 0x616E00;

    private static final AnBuiltinModules INSTANCE = new AnBuiltinModules();

    private final Map<String, AnBuiltinModule> modules = new LinkedHashMap<>();
    private final Map<Integer, AnBuiltinModule> byRequestCode = new HashMap<>();

    private Activity host;
    private int nextRequestCode = FIRST_REQUEST_CODE;

    private AnBuiltinModules() {}

    /**
     * Called once, before creating the {@link AnRuntime}: the core builds one native module per name
     * when the engine starts, and anything arriving after that would no longer get in.
     */
    public static void install(Activity host) {
        INSTANCE.host = host;
        AnRuntime.setBuiltinModules(INSTANCE);
        if (!INSTANCE.modules.isEmpty()) {
            // The Activity was recreated —a rotation, a configuration change,
            // the app coming back from the background— and this registry is not:
            // it belongs to the process, like the plugins'. What has to be
            // renewed is the screen, because the one the modules were holding is
            // gone and presenting anything from it would do nothing at all.
            for (AnBuiltinModule module : INSTANCE.modules.values()) {
                module.attach(host);
            }
            return;
        }
        INSTANCE.register("files", new AnFiles());
        INSTANCE.register("share", new AnShare());
    }

    private void register(String name, AnBuiltinModule module) {
        if (modules.containsKey(name)) {
            Log.w(TAG, "the built-in module " + name + " was installed twice");
            return;
        }
        modules.put(name, module);
        module.attach(host);
    }

    /**
     * Hands a module its own request code, so that {@link #onActivityResult} can find it again. It
     * is called from {@code attach} and the number never changes for the life of the process.
     */
    public static int reserveRequestCode(AnBuiltinModule module) {
        int code = INSTANCE.nextRequestCode++;
        INSTANCE.byRequestCode.put(code, module);
        return code;
    }

    /** Rust calls in here from the UI thread, inside the frame. */
    public void dispatch(long id, String module, String method, String args) {
        AnBuiltinCall respond = new AnBuiltinCall(id);
        AnBuiltinModule implementation = modules.get(module);
        if (implementation == null) {
            // This should not be able to happen: the names come from BUILTIN_MODULES and they
            // are the same ones registered above. If it does happen it is said, rather than
            // leaving the promise hanging.
            respond.reject(
                    "the built-in module "
                            + module
                            + " is not installed in this shell; it is in BUILTIN_MODULES and"
                            + " nobody registered it");
            return;
        }
        JSONObject parsed;
        try {
            parsed = new JSONObject(args == null ? "{}" : args);
        } catch (Exception error) {
            // JS can send anything, not only an object. It is handed over empty rather than
            // blowing up: it is the same thing the iOS shell does.
            parsed = new JSONObject();
        }
        try {
            implementation.call(method, parsed, respond);
        } catch (RuntimeException error) {
            // An exception from a module cannot bring the frame down, but it cannot disappear
            // either: it becomes the rejection of that promise.
            respond.reject(module + "." + method + " threw: " + error);
        }
    }

    /** Forwarded by {@link MainActivity}. Returns whether it belonged to a built-in module. */
    public static boolean onActivityResult(int requestCode, int resultCode, Intent data) {
        AnBuiltinModule module = INSTANCE.byRequestCode.get(requestCode);
        if (module == null) {
            return false;
        }
        module.onActivityResult(resultCode, data);
        return true;
    }
}

File: shells/shared/AnBuiltinModules.swift
import Foundation

#if canImport(UIKit) && !os(watchOS)
    import UIKit
#elseif os(macOS)
    import AppKit
#endif

/// The answer to a built-in module call.
///
/// It is `AnPluginCall` under another name and over another mailbox, and it is
/// separate for the same reason the mailboxes are: a host with no plugin
/// registry —macOS, the watch— still has the built-ins, and neither shell has an
/// `AnPluginCall` to borrow.
///
/// It can be answered on the spot or kept and answered later: that is what
/// separates asking the network monitor from putting a share sheet on the screen
/// and waiting for somebody to pick something. What cannot be done is not
/// answering, so every error path has to end in `reject`.
final class AnBuiltinCall {
    private let id: UInt64
    private let lock = NSLock()
    private var answered = false

    init(id: UInt64) {
        self.id = id
    }

    /// For a method that returns nothing.
    func resolve() {
        send(json: "null")
    }

    func resolve(_ text: String) {
        send(json: Self.encode(text))
    }

    func resolve(_ value: Bool) {
        send(json: value ? "true" : "false")
    }

    func resolve(_ value: Double) {
        send(json: Self.encode(value))
    }

    func resolve(_ object: [String: Any]) {
        send(json: Self.encode(object))
    }

    func resolve(_ array: [Any]) {
        send(json: Self.encode(array))
    }

    /// For a method that answers "there was nothing", which is not the same as
    /// failing: a picker somebody dismissed, a share sheet somebody cancelled.
    func resolveNull() {
        send(json: "null")
    }

    func reject(_ message: String) {
        lock.lock()
        let first = !answered
        answered = true
        lock.unlock()
        guard first else { return }
        _ = an_builtin_reject(id, message)
    }

    private func send(json: String) {
        lock.lock()
        let first = !answered
        answered = true
        lock.unlock()
        guard first else {
            NSLog("angular-native: a built-in module answered the same call twice")
            return
        }
        _ = an_builtin_resolve(id, json)
    }

    private static func encode(_ value: Any) -> String {
        guard
            let data = try? JSONSerialization.data(
                withJSONObject: value, options: [.fragmentsAllowed]),
            let text = String(data: data, encoding: .utf8)
        else {
            NSLog("angular-native: a built-in module returned something that is not JSON")
            return "null"
        }
        return text
    }
}

/// What a built-in module implements. The name it answers to is in
/// `AnBuiltinModules.install`, and in `BUILTIN_MODULES` on the Rust side.
protocol AnBuiltinModule: AnyObject {
    /// Called once, at install time, before any call can arrive.
    func start()

    /// `args` is what JS sent, already decoded. If it sent something that is
    /// not an object, this arrives empty. A method that does not exist has to
    /// be rejected, not ignored.
    func call(_ method: String, _ args: [String: Any], _ respond: AnBuiltinCall)
}

extension AnBuiltinModule {
    func start() {}
}

/// Where a module puts something on the screen.
///
/// A share sheet and a file picker need somewhere to be presented from, and
/// that somewhere is not the same object on every Apple platform: UIKit
/// presents from a view controller, AppKit hangs a picker off a view, and
/// watchOS presents neither of the two because it has neither. It is kept here,
/// weakly, so a module does not have to be handed it and does not keep the
/// screen alive.
enum AnBuiltinHost {
    #if canImport(UIKit) && !os(watchOS)
        static weak var viewController: UIViewController?
    #elseif os(macOS)
        static weak var view: NSView?
    #endif
}

/// The built-in modules of an Apple shell.
///
/// It is the twin of `AnPluginRegistry` with one difference that matters: who
/// is inside is not read from anybody's `package.json`. These four ship with the
/// framework and every host has them, so the list is written here and in
/// `BUILTIN_MODULES` on the Rust side, and `scripts/check-builtins.sh` reads
/// both.
enum AnBuiltinModules {
    private static var modules: [String: AnBuiltinModule] = [:]

    /// Called once, before the runtime is created: the core builds one native
    /// module per name when the engine starts, and anything arriving later
    /// would no longer get in.
    static func install() {
        an_builtin_set_dispatch { id, module, method, args in
            AnBuiltinModules.dispatch(
                id: id,
                module: AnBuiltinModules.text(module),
                method: AnBuiltinModules.text(method),
                args: AnBuiltinModules.text(args))
        }
        register("files", AnBuiltinFiles())
        register("share", AnBuiltinShare())
    }

    private static func register(_ name: String, _ module: AnBuiltinModule) {
        guard modules[name] == nil else {
            NSLog("angular-native: the built-in module \(name) was installed twice")
            return
        }
        modules[name] = module
        module.start()
    }

    /// Rust calls in here from the main thread, inside the frame.
    private static func dispatch(id: UInt64, module: String, method: String, args: String) {
        let respond = AnBuiltinCall(id: id)
        guard let implementation = modules[module] else {
            // This should not be able to happen: the names come from
            // `BUILTIN_MODULES` and they are the same ones registered above. If
            // it does happen, it is said rather than leaving the promise
            // hanging.
            respond.reject(
                "the built-in module \(module) is not installed in this shell; "
                    + "it is in BUILTIN_MODULES and nobody registered it")
            return
        }
        let decoded = try? JSONSerialization.jsonObject(
            with: Data(args.utf8), options: [.fragmentsAllowed])
        implementation.call(method, decoded as? [String: Any] ?? [:], respond)
    }

    private static func text(_ pointer: UnsafePointer<CChar>?) -> String {
        guard let pointer else { return "" }
        return String(cString: pointer)
    }
}
